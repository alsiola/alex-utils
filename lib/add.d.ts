import { CurriedA2Function } from "./types";
export declare const add: CurriedA2Function<number, number, number>;
