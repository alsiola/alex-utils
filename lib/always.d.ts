/**
 * Creates a function that always returns a
 * @param a The value to return from returned function
 */
export declare const always: <T>(a: T) => () => T;
