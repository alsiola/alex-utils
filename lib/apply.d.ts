import { AnyFunction } from "./types";
export declare const apply: <T>(f: AnyFunction<T>, thisArg?: any) => (...a: any[]) => T;
