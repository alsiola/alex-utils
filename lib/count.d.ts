export declare const count: (a: any[]) => number;
