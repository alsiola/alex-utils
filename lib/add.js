"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var curry_1 = require("./curry");
exports.add = curry_1.curry2(function (a, b) { return a + b; });
