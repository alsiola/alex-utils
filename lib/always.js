"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Creates a function that always returns a
 * @param a The value to return from returned function
 */
exports.always = function (a) { return function () { return a; }; };
