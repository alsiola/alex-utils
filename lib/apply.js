"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = function (f, thisArg) { return function () {
    var a = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        a[_i] = arguments[_i];
    }
    return f.apply(thisArg, a);
}; };
