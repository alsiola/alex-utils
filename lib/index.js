"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var add_1 = require("./add");
exports.add = add_1.add;
var always_1 = require("./always");
exports.always = always_1.always;
var count_1 = require("./count");
exports.count = count_1.count;
var curry_1 = require("./curry");
exports.curryV = curry_1.curryV;
exports.curry2 = curry_1.curry2;
var delay_1 = require("./delay");
exports.delay = delay_1.delay;
var divide_1 = require("./divide");
exports.divide = divide_1.divide;
var flip_1 = require("./flip");
exports.flip = flip_1.flip;
exports.curryFlip = flip_1.curryFlip;
var fork_1 = require("./fork");
exports.fork = fork_1.fork;
exports.forkAsync = fork_1.forkAsync;
var identity_1 = require("./identity");
exports.identity = identity_1.identity;
var makeRecord_1 = require("./makeRecord");
exports.makeRecord = makeRecord_1.makeRecord;
var map_1 = require("./map");
exports.map = map_1.map;
exports.mapAsync = map_1.mapAsync;
exports.mapObject = map_1.mapObject;
exports.mapObjectAsync = map_1.mapObjectAsync;
var pipe_1 = require("./pipe");
exports.pipe = pipe_1.pipe;
exports.pipeAsync = pipe_1.pipeAsync;
var prop_1 = require("./prop");
exports.prop = prop_1.prop;
var reduce_1 = require("./reduce");
exports.reduce = reduce_1.reduce;
exports.reduceAsync = reduce_1.reduceAsync;
var sum_1 = require("./sum");
exports.sum = sum_1.sum;
var tap_1 = require("./tap");
exports.tap = tap_1.tap;
