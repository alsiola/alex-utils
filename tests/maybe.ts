import { propM, Nothing, Just, lift, Maybe } from "../lib";

describe("maybe", () => {
    it("map applies fn to wrapped value in Just", () => {
        const input = Just.from(5);

        const actual = input.map(x => x + 1);

        expect(actual).toBeInstanceOf(Just);
        expect(actual.withDefault(100)).toBe(6);
    });

    it("map applies fn to wrapped value in Nothing", () => {
        const input: Maybe<number> = Nothing.from();

        const actual = input.map(x => x + 1);

        expect(actual).toBeInstanceOf(Nothing);
        expect(actual.withDefault(100)).toBe(100);
    });

    it("apply passes value to wrapped fn", () => {
        const input = Just.from((x: number) => x + 1);

        const actual = input.apply(5);

        expect(actual).toBeInstanceOf(Just);
        expect(actual.withDefault(500)).toBe(6);
    });

    it("apply throws if wrapped value is not function", () => {
        const input = Just.from(5);

        expect(() => input.apply(7)).toThrowErrorMatchingSnapshot();
    });
});

describe("propM", () => {
    it("propM returns just value if present", () => {
        const input = { a: 1 };

        const actual = propM("a")(input);

        const expected = 1;

        expect(actual).toBeInstanceOf(Just);
        expect(actual.withDefault(5)).toBe(expected);
    });

    it("propM returns nothing if not present", () => {
        const input = { a: 1 };

        const actual = propM("b" as any)(input);

        const expected = 100;

        expect(actual).toBeInstanceOf(Nothing);
    });
});

describe("lift", () => {
    it("converts arity-1 function to work on maybe", () => {
        const f = lift((x: number) => x + 1);

        const input = Just.from(5);

        const actual = f(input);

        expect(actual).toBeInstanceOf(Just);
        expect(actual.withDefault(500)).toBe(6);
    });
});
