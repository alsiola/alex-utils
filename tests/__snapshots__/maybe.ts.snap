// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maybe apply throws if wrapped value is not function 1`] = `"Apply is only callable if wrapped value is a function"`;
